From: Alessandro Astone <alessandro.astone@canonical.com>
Date: Thu, 25 Jun 2026 17:19:23 +0200
Subject: gsk: Account for render offset when snapping glyphs

Glyph snapping and subpixel cache flags need to be computed in the same
coordinate space that is eventually rendered. Commit 96f65b0e moved
render-pass offset handling into the shader ops, but left glyph origins
snapped before that offset was applied.

Include self->offset while snapping the glyph origin and deriving glyph
lookup flags, then subtract it again before passing bounds to the shader
ops.

Fixes: 96f65b0ef5 ("gpu: Refactor glyphs code")
Bug: https://gitlab.gnome.org/GNOME/gtk/-/work_items/8194
Forwarded: https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/10091
---
 gsk/gpu/gskgpunodeprocessor.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/gsk/gpu/gskgpunodeprocessor.c b/gsk/gpu/gskgpunodeprocessor.c
index a3242cb..ac9428d 100644
--- a/gsk/gpu/gskgpunodeprocessor.c
+++ b/gsk/gpu/gskgpunodeprocessor.c
@@ -2658,14 +2658,14 @@ gsk_gpu_node_processor_add_glyph_node (GskGpuRenderPass *self,
       graphene_point_t glyph_offset, glyph_origin;
       GskGpuGlyphLookupFlags flags;
 
-      glyph_origin = GRAPHENE_POINT_INIT (offset.x + glyphs[i].geometry.x_offset / pango_scale,
-                                          offset.y + glyphs[i].geometry.y_offset / pango_scale);
+      glyph_origin = GRAPHENE_POINT_INIT (offset.x + glyphs[i].geometry.x_offset / pango_scale + self->offset.x,
+                                          offset.y + glyphs[i].geometry.y_offset / pango_scale + self->offset.y);
 
       glyph_origin.x = floorf (glyph_origin.x * align_scale_x + 0.5f);
       glyph_origin.y = floorf (glyph_origin.y * align_scale_y + 0.5f);
       flags = (((int) glyph_origin.x & 3) | (((int) glyph_origin.y & 3) << 2)) & flags_mask;
-      glyph_origin.x /= align_scale_x;
-      glyph_origin.y /= align_scale_y;
+      glyph_origin.x = glyph_origin.x / align_scale_x - self->offset.x;
+      glyph_origin.y = glyph_origin.y / align_scale_y - self->offset.y;
 
       image = gsk_gpu_cached_glyph_lookup (cache,
                                            self->frame,
